### PR TITLE
Record subtyping

### PR DESCRIPTION
**Pull request by luk-f-a**
First attempt at getting record (width) subtyping, following suggestion at https://github.com/numba/numba/issues/5554#issuecomment-613489530

### CODE REVIEW

**Comment by luk-f-a** [comment]
Status: I added one test in https://github.com/luk-f-a/numba/blob/3ee37f7f981d688c822d98cac98ceba9e0cd002d/numba/tests/test_record_dtype.py#L1081 but it's currently failing with 

`TypeError: No matching definition for argument type(s) Record(a[type=float64;offset=0],b[type=float64;offset=8];16;False)`

**Comment by sklam** [comment]
Try this patch: https://gist.github.com/sklam/a1a575d609d54f386b32abc809aa19ba

There's a bug in the dispatcher logic in C. It fails to disable `exact_match_required` when compilation is disabled.

**Comment by luk-f-a** [comment]
thanks a lot @sklam, it works well. I'll write more tests tomorrow to cover as many edge cases as possible. After that, I'll create a second PR for the subtyping of function types (2nd part of the proposal).

**Comment by sklam** [comment]
There were a few review/consideration notes from today's developer meeting discussions on this PR:

* Is it safe in regard to optimization passes? i.e. branch-pruning
* It seems to be sensitive to the compilation order. Does it matter?

**Comment by luk-f-a** [comment]
thanks for the review @sklam ! I'll go point by point between today and tomorrow, but I'll start with the first point, the behaviour of arrays.

If I understood your example, you would expect an array of record subtypes, to be itself a subtype of the array of main record types.
Wouldn't that be dangerous? Arrays allow `append` which would then lead to an array of mixed types that would have an undefined, probably wrong behaviour. 
Please excuse me if I got it wrong, because I am not a CS guy, but I thought read/write arrays are supposed to be invariant (rather than covariant), which is why the behaviour in your example to me looks like the safe thing to do.

Tuples are read-only and can be covariant, and tuples-of-records currently work like this on this branch

```python
from numba.core import types
rec1 = np.array([1], dtype=[('a', 'f8')])[0]
rec2 = np.array([(2,3)], dtype=[('a', 'f8'), ('b', 'f8')])[0]

tup1 = (rec1,)
tup2 = (rec2,)

@njit(types.float64(typeof(tup1)))
def foo(rec):
    return rec[0]['a']

print(foo(tup2)) # works
```

File: numba/tests/test_record_dtype.py [hunk 1]
```diff
@@ -1077,5 +1077,104 @@ def test_error_w_invalid_field(self):
...
+        c_dtype = np.dtype([('c', 'f8')])
+        c_rec1 = np.array([1], dtype=c_dtype)[0]
+
+        @njit
+        def foo(rec):
+            return rec['c']
+
+        foo(c_rec1)
+        foo.disable_compile()
+        with self.assertRaises(TypeError):
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/tests/test_record_dtype.py]: good point, fixed it in all 3 places where it happens.

File: numba/tests/test_record_dtype.py [hunk 2]
```diff
@@ -1077,5 +1077,104 @@ def test_error_w_invalid_field(self):
...
+            foo(self.a_rec1)
+
+    def test_no_subtyping2(self):
+        """
+        test that conversion rules don't allow smaller records as subtypes
+        """
+        jit_fc = njit(self.func)
+        jit_fc(self.ab_rec1)
+        jit_fc.disable_compile()
+        with self.assertRaises(TypeError):
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/tests/test_record_dtype.py]: fixed

File: numba/tests/test_record_dtype.py [hunk 3]
```diff
@@ -1077,5 +1077,104 @@ def test_error_w_invalid_field(self):
...
+    def test_no_subtyping3(self):
+        """
+        test that conversion rules don't allow records with fields with same
+        name but incompatible type
+        """
+        other_a_rec = np.array(['a'], dtype=np.dtype([('a', 'U25')]))[0]
+        jit_fc = njit(self.func)
+        jit_fc(self.a_rec1)
+        jit_fc.disable_compile()
+        with self.assertRaises(TypeError):
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/tests/test_record_dtype.py]: fixed

**Comment by luk-f-a** [comment]
@sklam I've addressed all the points, whenever you have time it's ready for a second round.

**Comment by sklam** [comment]
@luk-f-a, with append-able record arrays, the situation is unsafe. But it would be safe for readonly operations or even inplace mutations on records arrays. If it does not support readonly arrays, the enhancement would seem overly limited, as the compilation overhead on functions with loops over arrays would be far greater than scalar functions. However, I can see important use-cases where you want to expose the scalar functions to an external system that provides the looping.

Right now, I am just wondering if this feature can have extensions to containers or is it a one-off feature for numpy record scalars and tuples of records.

**Comment by luk-f-a** [comment]
@sklam you're right, I hadn't thought of read only arrays. The good news is that your example works when using read only arrays

```python
import numpy as np

from numba import njit

rec1 = np.empty(5, dtype=[('a', 'f8')])

rec1['a'] = 1
rec1.flags.writeable = False
rec2 = np.empty(5, dtype=[('a', 'f8'), ('b', 'i8')])
rec2['a'] = 2
rec2['b'] = 3
rec2.flags.writeable = False

print(rec1)
print(rec2)

@njit
def foo(rec, flag=None):
    n = 0
    for a in rec[::2]:
        n += a['a']
    return n


print(foo(rec1))
foo.disable_compile()
print(foo.nopython_signatures)

print(foo(rec1))
print(foo(rec2))
```

The great thing is that I didn't have to do anything. The rules for subtyping containers are already coded in the container types themselves. So I just added the rules for subtyping records, and tuples and read-only arrays just worked.

Regarding in-place mutation of subtyped arrays, wouldn't that require a new array type that allows in-place mutation but does not allow appends? Something like `Array` type class having a property `appendable` that can be set to `False` independently of `mutable`.
Regarding typed.List, if it implemented a `can_convert_to` method, I don't see why it wouldn't work with records, assuming that typed.List always stores pointers and not values.

> I can see important use-cases where you want to expose the scalar functions to an external system that provides the looping.

That's exactly what I do :-)

**Comment by sklam** [comment]
Given that the Record subtyping feature seems overly limited currently and there are plans to extend it for readonly Array, which may open up new problems, we should mark this feature as experimental for now. I'd suggest this PR to add:

- documentation of the feature with note saying it is experimental
- a warning message when subtype conversion is happening

**Comment by luk-f-a** [comment]
@sklam , it makes sense, a new behaviour without warning would be confusing. I've added both documentation and a warning.

**Comment by sklam** [comment]
@luk-f-a, can you add https://github.com/sklam/numba/commit/8de4467c62453a93e8344a3e71e2a67af3f3ac59 so that the warnings are tested?

**Review by sklam — APPROVED** [review]

**Comment by luk-f-a** [comment]
@sklam no problem, I'll do it later today.
I'm happy this is approved but also a bit surprised. After last month's open dev meeting, I thought you would have wanted subtyping to work via an extension mechanism (probably a new one to be defined), and not necessarily be part of main numba.

**Comment by luk-f-a** [comment]
@sklam I don't fully understand the intention in the patch you provided. By turning all warnings into exception all tests fail now fail. Do you think I should change all tests to catch the warning? Or should I add 1 new test that checks that the code raises an ExperimentalWarning? The changes you made imply the first option, but warnings and exceptions are usually tested in a separate test.

Also, instead of using `warnings.simplefilter("error"`, wouldn't it be better to use `assertWarns`?

**Comment by sklam** [comment]
> I'm happy this is approved but also a bit surprised. After last month's open dev meeting, I thought you would have wanted subtyping to work via an extension mechanism (probably a new one to be defined), and not necessarily be part of main numba.

It will be difficult to implement features that alters behavior of existing types as an extension. Also, I find this feature to be well behaved and have a limited scope that I can justify it being an experimental feature on mainline.

> @sklam I don't fully understand the intention in the patch you provided. By turning all warnings into exception all tests fail. Do you think I should change all tests to catch the warning? Or should I add 1 new test that checks that the code raises an ExperimentalWarning? The changes you made imply the first option, but warnings and exceptions are usually tested in a separate test.

Sorry for the confusion. I thought I added `assertWarns` to all the places that needed it. The reason to turn warnings into exception is that I wanted to test exactly where the warnings are raised. However, once I tried to add all the necessary assertWarn, it revealed something I missed---The casting result is being cached and the warning will not occur for the same pair of record types. This means there is no easy way to `assertWarns` unless all tests are using unique record types.

I think it is best to revert https://github.com/numba/numba/pull/5560/commits/51b72272b49bdd8713b6f0792207dcb7c9351404.

**Review by sklam — APPROVED** [review]
Thanks for the patch

**Pull request merged**